modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java

```diff
@@ -357,6 +381,61 @@ public void testBulkRejectionsRetryAndFailAnyway() throws Exception {
         assertEquals(retryAttempts, task.getStatus().getRetries());
     }
 
+    public void testPerfectlyThrottledBatchTime() {
+        DummyAbstractAsyncBulkByScrollAction action = new DummyAbstractAsyncBulkByScrollAction();
+        mainRequest.setRequestsPerSecond(0);
+        assertThat((double) action.perfectlyThrottledBatchTime(randomInt()), closeTo(0f, 0f));
+
+        int total = between(0, 1000000);
+        task.rethrottle(1);
+        assertThat((double) action.perfectlyThrottledBatchTime(total),
+                closeTo(TimeUnit.SECONDS.toNanos(total), TimeUnit.SECONDS.toNanos(1)));
+    }
+
+    public void testScrollDelay() throws Exception {
+        /*
+         * Replace the thread pool with one that will save the delay sent for the command. We'll use that to check that we used a proper
+         * delay for throttling.
+         */
+        final AtomicReference<TimeValue> capturedDelay = new AtomicReference<>();
+        final AtomicReference<Runnable> capturedCommand = new AtomicReference<>();
+        threadPool.shutdown();
+        threadPool = new ThreadPool(getTestName()) {
+            @Override
+            public ScheduledFuture<?> schedule(TimeValue delay, String name, Runnable command) {
+                capturedDelay.set(delay);
+                capturedCommand.set(command);
+                return null;
+            }
+        };
+
+        DummyAbstractAsyncBulkByScrollAction action = new DummyAbstractAsyncBulkByScrollAction();
+        action.setScroll(scrollId());
+
+        // We'd like to get about 1 request a second
+        task.rethrottle(1f);
+        // Make the last scroll look nearly instant
+        action.setLastBatchStartTime(System.nanoTime());
+        // The last batch had 100 documents
+        action.startNextScroll(100);
+
+        // So the next request is going to have to wait an extra 100 seconds or so (base was 10, so 110ish)
+        assertThat(client.lastScroll.get().request.scroll().keepAlive().seconds(), either(equalTo(110L)).or(equalTo(109L)));
+
+        // Now we can simulate a response and check the delay that we used for the task
+        InternalSearchHit hit = new InternalSearchHit(0, "id", new Text("type"), Collections.<String, SearchHitField>emptyMap());
+        InternalSearchHits hits = new InternalSearchHits(new InternalSearchHit[] { hit }, 0, 0);
+        InternalSearchResponse searchResponse = new InternalSearchResponse(hits, null, null, null, false, false);
+        client.lastScroll.get().listener.onResponse(new SearchResponse(searchResponse, scrollId(), 5, 4, randomLong(), null));
+
+        // The delay is still 100ish seconds because there hasn't been much time between when we requested the bulk and when we got it.
+        assertThat(capturedDelay.get().seconds(), either(equalTo(100L)).or(equalTo(99L)));
```

Comment:
It hasn't yet and I think it is pretty safe - the time is between the `setLastBatchStartTime` and `startNextScroll` and those values are in seconds. I think floating point rounding errors are larger. But I could factor the time check in startNextScroll in such a way that the test can spoof it and we don't need any time.